[Yan] 스프링 카페 2단계 - 글 쓰기 기능 구현

KTH-96 opened the pull request:
안녕하세요 Yan입니다 :)
2단계 글쓰기 구현및 회원정보 수정 구현입니다.
리뷰항상 감사드립니다 ㅎㅎ

KTH-96 committed:
feat : 게시물 필요한 변수와 저장하는 로직

현 상황에서 dto는 필요없다고 판단하여 만들지 않았습니다

KTH-96 committed:
feat : 게시물 서비스 로직 및 컨트롤러 생성

인덱스 페이지에서 저장한 게시물만 볼수있도록 인덱스 컨트롤러 변경했습니다

KTH-96 committed:
feat : 인덱스 페이지 변경

인덱스 페이지에 타임리프 문법 적용했습니다

KTH-96 committed:
feat : 게시판 타임리프 문법 적용

KTH-96 committed:
feat : 게시글 등록시간 추가 및 게시글 상세페이지 템플릿 구현

레포지토리에서 게시글 정보를 저장할때 시간을 같이 저장해주었습니다

KTH-96 committed:
feat : 회원목록에서 회원 정보수정하는로직

User에 수정된 데이터 받을수있게 setter추가

KTH-96 committed:
feat : 회원정보 수정할 수 있도록 html추가

form을 복사해서 userForm생성
put매핑을 사용할수 있도록 properties에 내용추가

tmdgusya commented on src/main/java/com/kakao/cafe/domain/posts/Posts.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.time.LocalDateTime;
+
+public class Posts {
```
왜 복수형으로 하신건가요? 이유가 있으신건가요

KTH-96 replied:
아뇨 하다보니 그렇게 한거 같습니다 post로 변경했습니다 :)

tmdgusya commented on src/main/java/com/kakao/cafe/domain/posts/Posts.java:
```diff
@@ -0,0 +1,46 @@
+        this.localDateTime = localDateTime;
+    }
+
+    public Long getId() {
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public String getWriter() {
+        return writer;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public String getContents() {
+        return contents;
+    }
+
+    public LocalDateTime getLocalDateTime() {
+        return localDateTime;
+    }
```
`getter` 는 전부 필요한가요?

KTH-96 replied:
getter는 html파일에서 값을 가져오려면 사용해야해서 다 만들었습니다. 혹시 더 좋은 방법이있다면 힌트 부탁드리겠습니다! :)

tmdgusya replied:
Response DTO 를 만들어도 가능하겠죠 ㅎㅎ?

tmdgusya commented on src/main/java/com/kakao/cafe/domain/posts/PostsRepository.java:
```diff
@@ -0,0 +1,27 @@
+public class PostsRepository {
+
+    public static Map<Long, Posts> postsMap = new HashMap<>();
+    private static long seq = 0L;
```
`seq` 를 사용하신 이유는 무엇일까요 ㅎㅎ?

KTH-96 replied:
Post에서 글쓴이가 하나의 게시물만 작성하는게 아니라 여려개를 작성했을때에 글쓴이가 key값으로 가지기엔 부적절하다고 생각했어서 만들었습니다.

tmdgusya replied:
넵넵 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/domain/user/User.java:
```diff
@@ -37,4 +37,20 @@ public String getName() {
     public String getEmail() {
         return email;
     }
+
+    public void setUserId(String userId) {
+        this.userId = userId;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public void setEmail(String email) {
+        this.email = email;
+    }
```
`setter` 는 다 필요한가요?

tmdgusya commented on src/main/java/com/kakao/cafe/service/PostsService.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+
+    public void qnaRegistration(Posts posts) {
```
함수의 이름은 동사형인게 좋습니다 ㅎㅎ

KTH-96 replied:
감사합니다 수정하겠습니다 :)

tmdgusya commented on src/main/java/com/kakao/cafe/service/PostsService.java:
```diff
@@ -0,0 +1,31 @@
+        postsRepository.save(posts);
+    }
+
+    public List<Posts> qnaList() {
+        return postsRepository.findAll();
+    }
+
+    public Posts qnaShowContent(Long id) {
+        return postsRepository.findById(id);
+    }
```
요 2개 함수도 마찬가지입니다 ㅎㅎ

tmdgusya commented on src/main/java/com/kakao/cafe/web/PostsController.java:
```diff
@@ -0,0 +1,46 @@
+
+    @GetMapping("/form")
+    public String form() {
+        log.info("qna form");
```
로그를 좀 더 자세하게 찍어보는건 어떨까요?

tmdgusya commented on src/main/java/com/kakao/cafe/web/PostsController.java:
```diff
@@ -0,0 +1,46 @@
+    private final PostsService postsService;
+
+    @Autowired
+    public PostsController(PostsService postsService) {
```
`@AutoWired` 가 꼭 필요할까요?

KTH-96 replied:
생성자가 한게일때는 @AutoWired생략이 가능한거군요 감사합니다!!

tmdgusya commented on src/main/java/com/kakao/cafe/web/UserController.java:
```diff
@@ -48,4 +46,18 @@ public String profile(@PathVariable String userId, Model model) {
+    }
+
+    @PutMapping("/{id}/form")
+    public String update(@PathVariable Long id, User user){
```
어떻게 아무것도 붙여주지 않았는데 Json 을 `User` 로 바꿔줄 수 있는 걸까요? 이유를 혹시 아시나요?

KTH-96 replied:
Json정보를 받을때 보통 @RequestBody를 사용해서 받을 수 있다고 알고있습니다. 다만 @RequestBody를 생략해서 사용한다면 @ModelAttribute가 적용된다고 알고있어서 생략하여 만들었습니다.

tmdgusya commented on src/main/java/com/kakao/cafe/domain/user/UserRepository.java:
```diff
@@ -31,6 +31,15 @@ public User findByUserId(String userId) {
 
+    public User updateUser(Long id, User userUpdate) {
+        User user = userMap.get(id);
+        user.setUserId(userUpdate.getUserId());
+        user.setPassword(userUpdate.getPassword());
+        user.setName(userUpdate.getName());
+        user.setEmail(userUpdate.getEmail());
```
궁금한점이 하나 있는데 이걸 위해서 setter 를 전부 연거라면 setXXX 가 아니라 
뭔가 지금 로직에 맞는 이름으로 바꿔야 할거 같은데요 ㅎㅎ
`setter` 를 전부다 여는건 좋지 않은 습관입니다. 

`updateProfile` 이런걸로 바꿔보세요~

KTH-96 replied:
오 setter를 다 생성 할 필요가없군요 감사합니다 수정했습니다 :)

tmdgusya reviewed (changes requested):
미션 2단계 하시느라 고생이 많았습니다.

몇가지 언급해드릴 부분이 있는데요.
1. 함수의 이름은 왠만해서 동사형으로 적는다.
2. `getXXX`, `setXXX` 를 왠만해서 다 열지 않는다. -> 협력을 요청한다고 생각해야 합니다. 회원가입을 하는데 전부 다 수정할 수 있다고해서 속성 전부 setter 를 여는게 아니라, 회원가입에 대한 인터페이스를 제공하세요. 모든 속성에 setter 를 하는 것은 좋지 않은 습관입니다.
3. 무엇이 남이 읽기 좋은 코드일까 생각해보세요. 
4. 로그는 오류가 났거나, 어떤 일이 발생했을때 보려고 남기는 것이다. -> 어떻게 찍어야 처음 보는 사람 혹은 자신도 잘 찾을 수 있을지 고민해보세요.

고쳐야 할 부분들이 보여서 Request Changes 로 드립니다.
수정사항 올려주시면 머지 해드릴께요.

KTH-96 committed:
refeator : setter사용 최소화, 클래스명 및 함수명 변경, log 자세히 해보기

User setter들을 UpdateProfile함수 생성해서 setter제거
UserUpdateDto 생성
log 메세지 한글로 변경 및 클래스명 함수명 변경
